src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2376,44 +2426,175 @@ private void SelectVoice(string voice)
-                    Log.DebugFormat("Performing StepList step: {0}.", vStep);
-                    if (vStep.StartsWith("Action>") &&
-                        (Enum.TryParse<FunctionKeys>(vStep.Substring(7), out FunctionKeys result)))
+                    KeyValue commandKeyValue;
+                    if (commandList.First().StartsWith("Action>") &&
+                        (Enum.TryParse<FunctionKeys>(commandList.First().Substring(7), out FunctionKeys actionFunctionKey)))
                     {
-                        singleKeyValue = new KeyValue(result);
-                        HandleFunctionKeySelectionResult(singleKeyValue);
+                        Log.InfoFormat("StepListCommand triggering function key {0} press", actionFunctionKey);
+                        commandKeyValue = new KeyValue(actionFunctionKey);
+                        KeySelectionResult(commandKeyValue, multiKeySelection);
                     }
-                    else if (vStep.StartsWith("Keyboard>") || vStep.StartsWith("KeyboardAndReturn>"))
+                    else if (commandList.First().StartsWith("Keyboard>") || commandList.First().StartsWith("KeyboardAndReturn>"))
                     {
                         Enums.Keyboards keyboardEnum;
-                        bool vReturn = (vStep.StartsWith("KeyboardAndReturn>"));
-                        int vIndex = (vStep.StartsWith("KeyboardAndReturn>")) ? 18 : 9;
-                        if (System.Enum.TryParse(vStep.Substring(vIndex), out keyboardEnum))
+                        bool vReturn = (commandList.First().StartsWith("KeyboardAndReturn>"));
+                        int vIndex = vReturn ? 18 : 9;
+                        if (System.Enum.TryParse(commandList.First().Substring(vIndex), out keyboardEnum))
                         {
-                            ChangeKeyboardKeyValue keyValue = new ChangeKeyboardKeyValue(keyboardEnum, vReturn);
-                            ProcessChangeKeyboardKeyValue(keyValue);
+                            Log.InfoFormat("StepListCommand changing keyboard to {0}. Replace previous keyboard={1}", keyboardEnum, vReturn);
+                            commandKeyValue = new ChangeKeyboardKeyValue(keyboardEnum, vReturn);
                         }
                         else
                         {
-                            ChangeKeyboardKeyValue keyValue = new ChangeKeyboardKeyValue(vStep.Substring(vIndex), vReturn);
-                            ProcessChangeKeyboardKeyValue(keyValue);
+                            commandKeyValue = new ChangeKeyboardKeyValue(commandList.First().Substring(vIndex), vReturn);
+                            Log.InfoFormat("StepListCommand changing keyboard to string {0}. Replace previous keyboard={1}", commandKeyValue.String, vReturn);
                         }
+                        KeySelectionResult(commandKeyValue, multiKeySelection);
+                    }
+                    else if (commandList.First().StartsWith("KeyToggle>"))

```

Comment:
I no longer send Toggle to KeyboardOutputService. I make the determination here whether to send Press or Release.